From 60d1ea99d641a47211271dfdea34b37b2a639c8d Mon Sep 17 00:00:00 2001
From: Fabien Parent <fparent@baylibre.com>
Date: Thu, 18 Nov 2021 12:07:58 +0100
Subject: [PATCH] blk: allow to define the default block size as env var

When the block size cannot be determined in HW, a default value
of 512 bytes block size is used. For UFS on the MTK board we need
to support 4kB block size. Allow to set a different default value
through the environment.

Upstream-Status: Pending

Signed-off-by: Fabien Parent <fparent@baylibre.com>
---
 include/blkdev.h |  5 ++++-
 lib/blkdev.c     | 10 ++++++++++
 2 files changed, 14 insertions(+), 1 deletion(-)

diff --git a/include/blkdev.h b/include/blkdev.h
index 6cbecbb65f82..4a004742d1c1 100644
--- a/include/blkdev.h
+++ b/include/blkdev.h
@@ -20,7 +20,7 @@
 # include <sys/mkdev.h>		/* major and minor on Solaris */
 #endif
 
-#define DEFAULT_SECTOR_SIZE       512
+#define DEFAULT_SECTOR_SIZE       blkdev_get_default_sector_size()
 
 #ifdef __linux__
 /* very basic ioctls, should be available everywhere */
@@ -110,6 +110,9 @@ int blkdev_get_size(int fd, unsigned long long *bytes);
 /* get 512-byte sector count */
 int blkdev_get_sectors(int fd, unsigned long long *sectors);
 
+/* get the default sector size if it cannot be determined by the kernel */
+int blkdev_get_default_sector_size(void);
+
 /* get hardware sector size */
 int blkdev_get_sector_size(int fd, int *sector_size);
 
diff --git a/lib/blkdev.c b/lib/blkdev.c
index c22853ddcbb0..8fa19e2db136 100644
--- a/lib/blkdev.c
+++ b/lib/blkdev.c
@@ -184,6 +184,16 @@ blkdev_get_sectors(int fd, unsigned long long *sectors)
 	return -1;
 }
 
+int
+blkdev_get_default_sector_size(void)
+{
+	char *p = getenv("BLK_DEFAULT_SECTOR_SIZE");
+	if (!p)
+		return 512;
+
+	return atoi(p);
+}
+
 /*
  * Get logical sector size.
  *
-- 
2.34.0

